File: packages/bex_flutter/ios/Classes/BexFullSdkBridge.swift
import Foundation
import UIKit
import Flutter
import BKMExpressSDK

final class BexFullSdkBridge {
  static let shared = BexFullSdkBridge()

  private var initializationToken: BKMExpress.InitializationToken?
  private var initializeConfig: [String: Any] = [:]
  private var flowInProgress = false

  func initialize(_ config: [String: Any], result: @escaping FlutterResult) {
    guard
      let authToken = stringValue(config["authToken"]), !authToken.isEmpty,
      let merchantId = stringValue(config["merchantId"]), !merchantId.isEmpty,
      let merchantUserId = stringValue(config["merchantUserId"]), !merchantUserId.isEmpty,
      let gsmRaw = stringValue(config["gsmNo"]), !gsmRaw.isEmpty,
      let environment = stringValue(config["environment"]), !environment.isEmpty
    else {
      reject(result, code: "invalid_argument", message: "Missing required initialize fields.")
      return
    }

    let currencyCode = stringValue(config["currencyCode"]) ?? "TRY"
    let transactionType = mapTransactionType(stringValue(config["transactionType"]) ?? "sale")
    let installmentRaw = (config["installmentCount"] as? NSNumber)?.intValue ?? 1

    guard let installmentCount = BKMExpress.InstallmentCount(installmentRaw) else {
      reject(result, code: "invalid_argument", message: "installmentCount must be a positive integer.")
      return
    }

    let gsmNo: BKMExpress.GSMNO
    do {
      gsmNo = try BKMExpress.GSMNO(gsmRaw)
    } catch {
      reject(
        result,
        code: "invalid_argument",
        message: "gsmNo must be a valid Turkish mobile number (5XXXXXXXXX)."
      )
      return
    }

    initializeConfig = config

    Task { @MainActor in
      do {
        let token = try await BKMExpress.initialize(
          context: .init(
            authToken: authToken,
            merchantID: merchantId,
            merchantUserID: merchantUserId,
            gsmNo: gsmNo,
            currencyCode: currencyCode,
            transactionType: transactionType,
            installmentCount: installmentCount,
            mode: mapMode(environment)
          )
        )
        self.initializationToken = token
        result(["ok": true])
      } catch let failure as BKMExpress.Failure {
        self.reject(result, code: self.mapFailureCode(failure), message: failure.message)
      } catch {
        self.reject(result, code: "unknown", message: error.localizedDescription)
      }
    }
  }

  func pay(
    _ payment: [String: Any],
    options: [String: Any],
    result: @escaping FlutterResult
  ) {
    guard let token = initializationToken else {
      reject(result, code: "not_initialized", message: "Call BexFullSdk.initialize() before pay/selectCard.")
      return
    }

    guard !flowInProgress else {
      reject(result, code: "already_in_progress", message: "A BEX SDK flow is already in progress.")
      return
    }

    guard
      let amountNumber = payment["amount"] as? NSNumber,
      let orderId = stringValue(payment["orderId"]), !orderId.isEmpty,
      let securityRaw = stringValue(payment["security"]),
      let installmentRaw = payment["installmentCount"] as? NSNumber
    else {
      reject(result, code: "invalid_argument", message: "Invalid payment payload.")
      return
    }

    guard let installmentCount = BKMExpress.InstallmentCount(installmentRaw.intValue) else {
      reject(result, code: "invalid_argument", message: "installmentCount must be a positive integer.")
      return
    }

    guard let presenter = Self.topViewController() else {
      reject(result, code: "no_activity", message: "No UIViewController available to present BKM Express.")
      return
    }

    let transactionIdString = stringValue(payment["transactionId"])
    let transactionID = transactionIdString.flatMap(UUID.init(uuidString:)) ?? UUID()
    let transactionDate = parseDate(stringValue(payment["transactionDate"])) ?? Date()

    flowInProgress = true
    let style = mapStyle(stringValue(options["style"]) ?? "fullScreen")
    let theme = resolveTheme(options: options)

    let paymentData = BKMExpress.PaymentData(
      amount: Decimal(amountNumber.doubleValue),
      security: mapSecurity(securityRaw),
      installmentCount: installmentCount,
      transactionType: mapTransactionType(stringValue(payment["transactionType"]) ?? "sale"),
      transactionID: transactionID,
      transactionDate: transactionDate,
      orderID: orderId,
      successUrl: stringValue(payment["successUrl"]) ?? "",
      failUrl: stringValue(payment["failUrl"]) ?? ""
    )

    DispatchQueue.main.async {
      BKMExpress.pay(
        token: token,
        data: paymentData,
        theme: theme,
        style: style,
        on: presenter
      ) { [weak self] payResult in
        guard let self else { return }
        self.flowInProgress = false
        switch payResult {
        case .completed:
          result([
            "status": "completed",
            "transactionId": transactionID.uuidString,
            "amount": amountNumber.doubleValue,
          ])
        case .cancelled:
          result(["status": "cancelled"])
        case .failed(let failure):
          if failure.isCancelled || self.mapFailureCode(failure) == "cancelled" {
            result(["status": "cancelled"])
          } else {
            self.reject(result, code: self.mapFailureCode(failure), message: failure.message)
          }
        }
      }
    }
  }

  func selectCard(_ options: [String: Any], result: @escaping FlutterResult) {
    guard let token = initializationToken else {
      reject(result, code: "not_initialized", message: "Call BexFullSdk.initialize() before pay/selectCard.")
      return
    }

    guard !flowInProgress else {
      reject(result, code: "already_in_progress", message: "A BEX SDK flow is already in progress.")
      return
    }

    guard let presenter = Self.topViewController() else {
      reject(result, code: "no_activity", message: "No UIViewController available to present BKM Express.")
      return
    }

    flowInProgress = true
    let style = mapStyle(stringValue(options["style"]) ?? "fullScreen")
    let theme = resolveTheme(options: options)

    DispatchQueue.main.async {
      BKMExpress.selectCard(
        token: token,
        theme: theme,
        style: style,
        on: presenter
      ) { [weak self] cardResult in
        guard let self else { return }
        self.flowInProgress = false
        switch cardResult {
        case .selected(let card):
          result([
            "status": "selected",
            "card": self.mapCard(card),
          ])
        case .cancelled:
          result(["status": "cancelled"])
        case .failed(let failure):
          if failure.isCancelled || self.mapFailureCode(failure) == "cancelled" {
            result(["status": "cancelled"])
          } else {
            self.reject(result, code: self.mapFailureCode(failure), message: failure.message)
          }
        }
      }
    }
  }

  private func mapCard(_ card: BKMExpress.Card) -> [String: Any] {
    var payload: [String: Any] = [
      "cardId": card.id.rawValue,
      "maskCardNumber": card.maskedCardNumber,
      "cardAlias": card.alias,
      "binValue": card.bin,
      "bankInformation": [
        "cardType": card.bankInformation.cardType,
        "cardBrandType": card.bankInformation.cardScheme,
        "cardBrand": card.bankInformation.cardBrand,
        "bankShortName": card.bankInformation.bankShortName,
        "bankCode": card.bankInformation.bankCode,
        "cardScheme": card.bankInformation.cardScheme,
      ] as [String: Any],
      "active": true,
    ]

    if let imageURL = card.imageURL?.absoluteString {
      payload["imageUrl"] = imageURL
    }

    return payload
  }

  private func resolveTheme(options: [String: Any]) -> BKMExpress.Theme {
    var theme = BKMExpress.Theme()

    let themeDict = (options["theme"] as? [String: Any])
      ?? (initializeConfig["theme"] as? [String: Any])

    guard let themeDict else { return theme }

    if let colors = themeDict["colors"] as? [String: Any] {
      let tintHex = stringValue(colors["tint"]) ?? stringValue(colors["primary"])
      if let tintHex, let color = UIColor(hex: tintHex) {
        theme.tint = color
      }
    }

    if let capsule = themeDict["buttonCapsule"] as? Bool, capsule {
      theme.buttonCornerRadius = .capsule
    } else if let radius = themeDict["buttonCornerRadius"] as? NSNumber {
      theme.buttonCornerRadius = .custom(CGFloat(truncating: radius))
    }

    return theme
  }

  private func mapMode(_ value: String) -> BKMExpress.Mode {
    switch value.lowercased() {
    case "preprod":
      return .preprod
    case "prod":
      return .production
    default:
      return .test
    }
  }

  private func mapSecurity(_ value: String) -> BKMExpress.PaymentSecurity {
    switch value.lowercased() {
    case "tds":
      return .tds
    case "otp":
      return .otp
    default:
      return .none
    }
  }

  private func mapTransactionType(_ value: String) -> BKMExpress.TransactionType {
    switch value.lowercased() {
    case "preauth", "pre_auth":
      return .preAuth
    case "recurring":
      return .recurring
    default:
      return .sale
    }
  }

  private func mapStyle(_ value: String) -> BKMExpress.PresentationStyle {
    value.lowercased() == "sheet" ? .sheet : .fullScreen
  }

  private func mapFailureCode(_ failure: BKMExpress.Failure) -> String {
    switch failure.kind {
    case .session:
      return "session"
    case .encryption:
      return "encryption"
    case .server:
      return "server"
    case .cancelled:
      return "cancelled"
    @unknown default:
      return "unknown"
    }
  }

  private func parseDate(_ value: String?) -> Date? {
    guard let value, !value.isEmpty else { return nil }
    if let millis = Double(value) {
      return Date(timeIntervalSince1970: millis > 10_000_000_000 ? millis / 1000.0 : millis)
    }
    let formatter = ISO8601DateFormatter()
    return formatter.date(from: value)
  }

  private func stringValue(_ value: Any?) -> String? {
    if let string = value as? String {
      return string
    }
    if let number = value as? NSNumber {
      return number.stringValue
    }
    return nil
  }

  private func reject(
    _ result: @escaping FlutterResult,
    code: String,
    message: String,
    title: String? = nil,
    nativeCode: Int? = nil
  ) {
    var details: [String: Any] = [
      "code": code,
      "message": message,
    ]
    if let title { details["title"] = title }
    if let nativeCode { details["nativeCode"] = nativeCode }
    result(
      FlutterError(code: code, message: message, details: details)
    )
  }

  private static func topViewController(
    base: UIViewController? = {
      UIApplication.shared.connectedScenes
        .compactMap { $0 as? UIWindowScene }
        .flatMap { $0.windows }
        .first { $0.isKeyWindow }?
        .rootViewController
    }()
  ) -> UIViewController? {
    if let navigation = base as? UINavigationController {
      return topViewController(base: navigation.visibleViewController)
    }
    if let tab = base as? UITabBarController {
      return topViewController(base: tab.selectedViewController)
    }
    if let presented = base?.presentedViewController {
      return topViewController(base: presented)
    }
    return base
  }
}

private extension UIColor {
  convenience init?(hex: String) {
    var normalized = hex.trimmingCharacters(in: .whitespacesAndNewlines).uppercased()
    if normalized.hasPrefix("#") {
      normalized.removeFirst()
    }

    var value: UInt64 = 0
    guard Scanner(string: normalized).scanHexInt64(&value) else {
      return nil
    }

    switch normalized.count {
    case 6:
      self.init(
        red: CGFloat((value & 0xFF0000) >> 16) / 255,
        green: CGFloat((value & 0x00FF00) >> 8) / 255,
        blue: CGFloat(value & 0x0000FF) / 255,
        alpha: 1
      )
    case 8:
      self.init(
        red: CGFloat((value & 0xFF000000) >> 24) / 255,
        green: CGFloat((value & 0x00FF0000) >> 16) / 255,
        blue: CGFloat((value & 0x0000FF00) >> 8) / 255,
        alpha: CGFloat(value & 0x000000FF) / 255
      )
    default:
      return nil
    }
  }
}

File: packages/bex_flutter/android/src/main/kotlin/com/bkm/bexflutter/BexFullSdkBridge.kt
package com.bkm.bexflutter

import android.app.Activity
import android.content.Context
import android.graphics.Color
import com.bkm.mobil.sdk.api.BexEnvironment
import com.bkm.mobil.sdk.api.BexFullSdk
import com.bkm.mobil.sdk.api.BexFullSdkConfig
import com.bkm.mobil.sdk.api.BexFullSdkTheme
import com.bkm.mobil.sdk.api.BexSdkError
import com.bkm.mobil.sdk.api.CardSelectionResult
import com.bkm.mobil.sdk.api.PaymentCallback
import com.bkm.mobil.sdk.api.PaymentResult
import com.bkm.mobil.sdk.api.PaymentSecurity
import com.bkm.mobil.sdk.api.SdkInitParams
import com.bkm.mobil.sdk.api.SdkMode
import com.bkm.mobil.sdk.api.SdkPaymentInfo
import com.bkm.mobil.sdk.api.TransactionType
import io.flutter.plugin.common.MethodChannel
import java.util.UUID
import java.util.concurrent.atomic.AtomicBoolean

class BexFullSdkBridge {
  private var appContext: Context? = null
  private var activityProvider: (() -> Activity?)? = null
  private var initializeConfig: Map<String, Any?>? = null
  private val flowInProgress = AtomicBoolean(false)

  fun attach(context: Context, activityProvider: () -> Activity?) {
    appContext = context.applicationContext
    this.activityProvider = activityProvider
  }

  fun initialize(config: Map<String, Any?>, result: MethodChannel.Result) {
    try {
      requireString(config, "authToken")
      requireString(config, "merchantId")
      requireString(config, "merchantUserId")
      requireString(config, "gsmNo")
      requireString(config, "environment")
      initializeConfig = config
      result.success(mapOf("ok" to true))
    } catch (error: Exception) {
      rejectError(result, "invalid_argument", error.message ?: "Invalid initialize config.")
    }
  }

  fun pay(
    payment: Map<String, Any?>,
    options: Map<String, Any?>,
    result: MethodChannel.Result
  ) {
    startFlow(
      mode = SdkMode.PAYMENT,
      payment = payment,
      options = options,
      result = result
    )
  }

  fun selectCard(options: Map<String, Any?>, result: MethodChannel.Result) {
    @Suppress("UNCHECKED_CAST")
    val paymentFromOptions = options["payment"] as? Map<String, Any?>
    startFlow(
      mode = SdkMode.CARD_SELECTION_ONLY,
      payment = paymentFromOptions,
      options = options,
      result = result
    )
  }

  private fun startFlow(
    mode: SdkMode,
    payment: Map<String, Any?>?,
    options: Map<String, Any?>,
    result: MethodChannel.Result
  ) {
    val config = initializeConfig
    if (config == null) {
      rejectError(result, "not_initialized", "Call BexFullSdk.initialize() before pay/selectCard.")
      return
    }

    if (!flowInProgress.compareAndSet(false, true)) {
      rejectError(result, "already_in_progress", "A BEX SDK flow is already in progress.")
      return
    }

    val activity = activityProvider?.invoke()
    val style = optionalString(options, "style") ?: "fullScreen"
    if (style == "sheet" && activity == null) {
      flowInProgress.set(false)
      rejectError(
        result,
        "no_activity",
        "No current Activity available for bottom sheet presentation."
      )
      return
    }

    try {
      val initParams = buildInitParams(config, payment)
      val theme = resolveTheme(config, options)
      val troySonic = when (val value = config["troySonicSoundEnabled"]) {
        is Boolean -> value
        else -> true
      }

      val settled = AtomicBoolean(false)

      fun settleSuccess(map: Map<String, Any?>) {
        if (settled.compareAndSet(false, true)) {
          flowInProgress.set(false)
          result.success(map)
        }
      }

      fun settleReject(
        code: String,
        message: String,
        title: String? = null,
        nativeCode: Int? = null
      ) {
        if (settled.compareAndSet(false, true)) {
          flowInProgress.set(false)
          rejectError(result, code, message, title, nativeCode)
        }
      }

      val initContext = activity ?: appContext
      if (initContext == null) {
        flowInProgress.set(false)
        rejectError(result, "no_activity", "No Android context available.")
        return
      }

      BexFullSdk.init(
        context = initContext,
        initParams = initParams,
        callback = object : PaymentCallback {
          override fun onPaymentSuccess(paymentResult: PaymentResult) {
            val map = mutableMapOf<String, Any?>(
              "status" to "completed",
              "transactionId" to paymentResult.transactionId,
              "amount" to paymentResult.amount
            )
            if (paymentResult.cardNumber != null) {
              map["cardNumber"] = paymentResult.cardNumber
            }
            settleSuccess(map)
          }

          override fun onCardSelected(cardResult: CardSelectionResult) {
            val card = cardResult.selectedCard
            val bank = mapOf(
              "cardType" to card.bexBankInformation.cardType,
              "cardBrandType" to card.bexBankInformation.cardBrandType,
              "cardBrand" to card.bexBankInformation.cardBrand,
              "bankShortName" to card.bexBankInformation.bankShortName
            )

            val cardMap = mutableMapOf<String, Any?>(
              "cardId" to card.cardId,
              "maskCardNumber" to card.maskCardNumber,
              "bankInformation" to bank,
              "active" to card.active
            )
            card.cardAlias?.let { cardMap["cardAlias"] = it }
            card.binValue?.let { cardMap["binValue"] = it }
            card.imageUrl?.let { cardMap["imageUrl"] = it }

            settleSuccess(
              mapOf(
                "status" to "selected",
                "card" to cardMap
              )
            )
          }

          override fun onError(error: BexSdkError) {
            if (error is BexSdkError.Cancelled) {
              settleSuccess(mapOf("status" to "cancelled"))
              return
            }

            when (error) {
              is BexSdkError.Network -> settleReject("network", error.displayMessage)
              is BexSdkError.Unauthorized -> settleReject("unauthorized", error.displayMessage)
              is BexSdkError.Api -> settleReject(
                "api",
                error.displayMessage,
                error.title,
                error.code
              )
              is BexSdkError.Unknown -> settleReject("unknown", error.displayMessage)
              is BexSdkError.Cancelled -> settleReject("cancelled", error.displayMessage)
            }
          }
        },
        theme = theme,
        config = BexFullSdkConfig(
          troySonicSoundEnabled = troySonic,
          mode = mode
        )
      )

      if (style == "sheet") {
        BexFullSdk.showAsBottomSheet(activity as Activity)
      } else {
        BexFullSdk.start()
      }
    } catch (error: Exception) {
      flowInProgress.set(false)
      rejectError(result, "unknown", error.message ?: "Failed to start BEX SDK flow.")
    }
  }

  private fun buildInitParams(
    config: Map<String, Any?>,
    payment: Map<String, Any?>?
  ): SdkInitParams {
    val currencyDefault = mapString(config, "currencyCode") ?: "TRY"
    val paymentInfo = buildPaymentInfo(config, payment, currencyDefault)

    return SdkInitParams(
      token = requireString(config, "authToken"),
      merchantId = requireString(config, "merchantId"),
      transactionId = optionalString(payment, "transactionId") ?: UUID.randomUUID().toString(),
      gsmNo = requireString(config, "gsmNo"),
      merchantUserId = requireString(config, "merchantUserId"),
      paymentInfo = paymentInfo,
      environment = mapEnvironment(requireString(config, "environment"))
    )
  }

  private fun buildPaymentInfo(
    config: Map<String, Any?>,
    payment: Map<String, Any?>?,
    currencyDefault: String
  ): SdkPaymentInfo {
    val amount = when (val raw = payment?.get("amount")) {
      is Number -> raw.toDouble()
      else -> 100.0
    }
    val installmentCount = when (val raw = payment?.get("installmentCount")) {
      is Number -> raw.toInt()
      else -> (config["installmentCount"] as? Number)?.toInt() ?: 1
    }

    return SdkPaymentInfo(
      amount = amount,
      orderId = optionalString(payment, "orderId")
        ?: "DEMO-${System.currentTimeMillis()}",
      transactionDate = optionalString(payment, "transactionDate")
        ?: System.currentTimeMillis().toString(),
      paymentSecurity = mapPaymentSecurity(optionalString(payment, "security") ?: "none"),
      currency = optionalString(payment, "currency") ?: currencyDefault,
      installmentCount = installmentCount,
      transactionType = mapTransactionType(
        optionalString(payment, "transactionType")
          ?: mapString(config, "transactionType")
          ?: "sale"
      ),
      successUrl = optionalString(payment, "successUrl")
        ?: "https://trcuzdan-dev.bkmtest.com.tr/sdk/demo/success",
      failUrl = optionalString(payment, "failUrl")
        ?: "https://trcuzdan-dev.bkmtest.com.tr/sdk/demo/fail"
    )
  }

  @Suppress("UNCHECKED_CAST")
  private fun resolveTheme(
    config: Map<String, Any?>,
    options: Map<String, Any?>
  ): BexFullSdkTheme? {
    val themeMap = (options["theme"] as? Map<String, Any?>)
      ?: (config["theme"] as? Map<String, Any?>)
      ?: return null

    val colorsMap = themeMap["colors"] as? Map<String, Any?>
    val colors = colorsMap?.let {
      BexFullSdkTheme.Colors(
        primary = parseColor(mapString(it, "primary")),
        primaryVariant = parseColor(mapString(it, "primaryVariant")),
        background = parseColor(mapString(it, "background")),
        surface = parseColor(mapString(it, "surface")),
        textPrimary = parseColor(mapString(it, "textPrimary")),
        textSecondary = parseColor(mapString(it, "textSecondary")),
        textOnPrimary = parseColor(mapString(it, "textOnPrimary")),
        buttonPrimary = parseColor(mapString(it, "buttonPrimary")),
        buttonPrimaryText = parseColor(mapString(it, "buttonPrimaryText")),
        buttonSecondaryBorder = parseColor(mapString(it, "buttonSecondaryBorder")),
        buttonSecondaryText = parseColor(mapString(it, "buttonSecondaryText")),
        buttonDisabled = parseColor(mapString(it, "buttonDisabled")),
        success = parseColor(mapString(it, "success")),
        error = parseColor(mapString(it, "error")),
        warning = parseColor(mapString(it, "warning")),
        border = parseColor(mapString(it, "border")),
        divider = parseColor(mapString(it, "divider"))
      )
    }

    val shape = BexFullSdkTheme.Shape(
      buttonCornerRadius = (themeMap["buttonCornerRadius"] as? Number)?.toFloat(),
      buttonBorderWidth = (themeMap["buttonBorderWidth"] as? Number)?.toFloat()
    )

    return BexFullSdkTheme(colors = colors, shape = shape)
  }

  private fun mapEnvironment(value: String): BexEnvironment = when (value.lowercase()) {
    "dev" -> BexEnvironment.DEV
    "test" -> BexEnvironment.TEST
    "preprod" -> BexEnvironment.PREPROD
    "prod" -> BexEnvironment.PROD
    else -> throw IllegalArgumentException("Unsupported environment: $value")
  }

  private fun mapPaymentSecurity(value: String): PaymentSecurity = when (value.lowercase()) {
    "tds" -> PaymentSecurity.TDS
    "otp" -> PaymentSecurity.OTP
    "none" -> PaymentSecurity.NONE
    else -> throw IllegalArgumentException("Unsupported payment security: $value")
  }

  private fun mapTransactionType(value: String): TransactionType = when (value.lowercase()) {
    "sale" -> TransactionType.SALE
    "preauth", "pre_auth" -> TransactionType.PRE_AUTH
    "recurring" -> TransactionType.RECURRING
    else -> throw IllegalArgumentException("Unsupported transaction type: $value")
  }

  private fun parseColor(value: String?): Int? {
    if (value.isNullOrBlank()) return null
    val normalized = if (value.startsWith("#")) value else "#$value"
    return try {
      Color.parseColor(normalized)
    } catch (_: IllegalArgumentException) {
      if (normalized.length == 9) {
        val rgb = normalized.substring(1, 7)
        val alpha = normalized.substring(7, 9)
        Color.parseColor("#$alpha$rgb")
      } else {
        null
      }
    }
  }

  private fun requireString(map: Map<String, Any?>, key: String): String {
    val value = mapString(map, key)?.trim()
    require(!value.isNullOrEmpty()) { "$key is required." }
    return value
  }

  private fun optionalString(map: Map<String, Any?>?, key: String): String? {
    if (map == null) return null
    return mapString(map, key)
  }

  private fun mapString(map: Map<String, Any?>, key: String): String? {
    return map[key]?.toString()
  }

  private fun rejectError(
    result: MethodChannel.Result,
    code: String,
    message: String,
    title: String? = null,
    nativeCode: Int? = null
  ) {
    val details = mutableMapOf<String, Any?>(
      "code" to code,
      "message" to message
    )
    if (title != null) details["title"] = title
    if (nativeCode != null) details["nativeCode"] = nativeCode
    result.error(code, message, details)
  }
}
